servers/physics/body_sw.cpp

```diff
@@ -54,33 +58,61 @@ void BodySW::update_inertias() {
 
 		case PhysicsServer::BODY_MODE_RIGID: {
 
-			//update tensor for allshapes, not the best way but should be somehow OK. (inspired from bullet)
+			//update tensor for all shapes, not the best way but should be somehow OK. (inspired from bullet)
 			float total_area=0;
 
 			for (int i=0;i<get_shape_count();i++) {
 
-				total_area+=get_shape_aabb(i).get_area();
+				total_area+=get_shape_area(i);
+			}
+
+			// We have to recompute the center of mass
+			center_of_mass_local.zero();
+
+			for (int i=0; i<get_shape_count(); i++) {
+				float area=get_shape_area(i);
+
+				float mass = area * this->mass / total_area;
+
+				// NOTE: we assume that the shape origin is also its center of mass
+				center_of_mass_local += mass * get_shape_transform(i).origin;
 			}
 
-			Vector3 _inertia;
+			center_of_mass_local /= mass;
 
+			// Recompute the inertia tensor
+			Matrix3 inertia_tensor;
+			inertia_tensor.set_zero();
 
 			for (int i=0;i<get_shape_count();i++) {
 
 				const ShapeSW* shape=get_shape(i);
 
-				float area=get_shape_aabb(i).get_area();
+				float area=get_shape_area(i);
 
 				float mass = area * this->mass / total_area;
 
-				_inertia += shape->get_moment_of_inertia(mass) + mass * get_shape_transform(i).get_origin();
+				Matrix3 shape_inertia_tensor=shape->get_moment_of_inertia(mass).to_diagonal_matrix();
+				Transform shape_transform=get_shape_transform(i);
+				Matrix3 shape_basis = shape_transform.basis.orthonormalized();
+
+				// NOTE: we don't take the scale of collision shapes into account when computing the inertia tensor!
+				shape_inertia_tensor = shape_basis * shape_inertia_tensor * shape_basis.transposed();
+
+				Vector3 shape_origin = shape_transform.origin - center_of_mass_local;
+				Matrix3 identity3(1, 0, 0, 0, 1, 0, 0, 0, 1);
+				inertia_tensor += shape_inertia_tensor + (identity3*shape_origin.dot(shape_origin)-shape_origin.outer(shape_origin))*mass;
+
 
 			}
 
-			if (_inertia!=Vector3())
-				_inv_inertia=_inertia.inverse();
-			else
-				_inv_inertia=Vector3();
+			if (Math::abs(inertia_tensor.determinant()) > CMP_EPSILON) {
```

Comment:
This make no sense, the determinant should be allowed to be smaller and we can allow it to be zero as we are computing the inverse through a simple transpose. Will fix asap